src/main/java/org/folio/services/invoice/PoLinePaymentStatusUpdateService.java

```diff
@@ -109,6 +120,81 @@ public Future<Void> updatePoLinePaymentStatusToCancelInvoice(Invoice invoiceFrom
+    Map<String, List<InvoiceLine>> poLineIdInvoiceLinesMap = groupInvoiceLinesByPoLineId(invoiceLines);
+    return fetchPoLines(poLineIdInvoiceLinesMap, requestContext)
+      .map(this::updatePoLinesPaymentStatus)
+      .compose(poLines -> orderLineService.updateCompositePoLines(poLines, requestContext));
+  }
+
+  private Map<String, List<InvoiceLine>>  groupInvoiceLinesByPoLineId(List<InvoiceLine> invoiceLines) {
+    if (CollectionUtils.isEmpty(invoiceLines)) {
+      return Collections.emptyMap();
+    }
+    return invoiceLines
+      .stream()
+      .filter(invoiceLine -> StringUtils.isNotEmpty(invoiceLine.getPoLineId()))
+      .collect(Collectors.groupingBy(InvoiceLine::getPoLineId));
+  }
+
+  /**
+   * Retrieves PO Lines associated with invoice lines and calculates expected PO Line's payment status
+   * @param poLineIdsWithInvoiceLines map where key is PO Line id and value is list of associated invoice lines
+   * @return map where key is {@link CompositePoLine} and value is expected PO Line's payment status
+   */
+  private Future<Map<CompositePoLine, CompositePoLine.PaymentStatus>> fetchPoLines(Map<String, List<InvoiceLine>> poLineIdsWithInvoiceLines,
+    RequestContext requestContext) {
+    List<Future<CompositePoLine>> futures = poLineIdsWithInvoiceLines.keySet()
+      .stream()
+      .map(poLine -> orderLineService.getPoLine(poLine, requestContext))
+      .toList();
+
+    return collectResultsOnSuccess(futures)
+      .map(compositePoLines ->  compositePoLines.stream()
+        .collect(Collectors.toMap(poLine -> poLine, poLine -> getPoLinePaymentStatus(poLineIdsWithInvoiceLines.get(poLine.getId())))));
+  }
+
+  private List<CompositePoLine> updatePoLinesPaymentStatus(Map<CompositePoLine, CompositePoLine.PaymentStatus> compositePoLinesWithNewStatuses) {
+    return compositePoLinesWithNewStatuses
+      .keySet().stream()
+      .filter(compositePoLine -> isPaymentStatusUpdateRequired(compositePoLinesWithNewStatuses, compositePoLine))
+      .map(compositePoLine -> updatePaymentStatus(compositePoLinesWithNewStatuses, compositePoLine))
+      .toList();
+  }
+
+  @VisibleForTesting
+  boolean isPaymentStatusUpdateRequired(Map<CompositePoLine, CompositePoLine.PaymentStatus> compositePoLinesWithStatus,
+      CompositePoLine compositePoLine) {
+    CompositePoLine.PaymentStatus newPaymentStatus = compositePoLinesWithStatus.get(compositePoLine);
+    return (!newPaymentStatus.equals(compositePoLine.getPaymentStatus()) &&
+      !PO_LINE_PAYMENT_IGNORED_STATUSES.contains(compositePoLine.getPaymentStatus()));
+  }
+
+  private CompositePoLine updatePaymentStatus(Map<CompositePoLine, CompositePoLine.PaymentStatus> compositePoLinesWithStatus,
+      CompositePoLine compositePoLine) {
+    CompositePoLine.PaymentStatus newPaymentStatus = compositePoLinesWithStatus.get(compositePoLine);
+    compositePoLine.setPaymentStatus(newPaymentStatus);
+    return compositePoLine;
```

Comment:
Can use just a single line here:
```suggestion
    return compositePoLine.withPaymentStatus(compositePoLinesWithStatus.get(compositePoLine));
```